### PR TITLE
Handle composition events to show suggestion on web

### PR DESCRIPTION
**Pull request by n2**
Handle composition events to show IME suggestion on web, as described in #248. Sorry for being late, I have been busy in the past few days.

### CODE REVIEW

File: egui_web/src/lib.rs [hunk 1]
```diff
@@ -708,22 +708,25 @@ fn install_text_agent(runner_ref: &AppRunnerRef) -> Result<(), JsValue> {
         let input_clone = input.clone();
         let runner_ref = runner_ref.clone();
         let on_compositionend = Closure::wrap(Box::new(move |event: web_sys::CompositionEvent| {
-            match event.type_().as_ref() {
+            // let event_type = event.type_();
```
Comment (emilk) [inline egui_web/src/lib.rs:711]:
```suggestion
```

File: egui_web/src/lib.rs [hunk 2]
```diff
@@ -708,22 +708,25 @@ fn install_text_agent(runner_ref: &AppRunnerRef) -> Result<(), JsValue> {
...
-                    if let Some(text) = event.data() {
-                        let mut runner_lock = runner_ref.0.lock();
-                        runner_lock.input.raw.events.push(egui::Event::Text(text));
-                        runner_lock.needs_repaint.set_true();
-                    }
+                    event.data().map(egui::Event::CompositionEnd)
                 }
-                "compositionupdate" => {}
+                "compositionupdate" => event.data().map(egui::Event::CompositionUpdate),
                 _s => panic!("Unknown type"),
```
Comment (emilk) [inline egui_web/src/lib.rs:725]:
```suggestion
                _s => {
                    console_err(format!("Unknown composition event type: {:?}", event.type_().as_ref());
                    None
                }
```

**Review by emilk — CHANGES_REQUESTED** [review]
Thank you!

File: egui/src/widgets/text_edit.rs [hunk 3]
```diff
@@ -503,6 +506,41 @@ impl<'t> TextEdit<'t> {
...
+                    Event::CompositionStart => {
+                        state.has_candidate = true;
+                        None
+                    }
+
+                    Event::CompositionUpdate(text_mark) => {
+                        if !text_mark.is_empty()
+                            && text_mark != "\n"
+                            && text_mark != "\r"
+                            && state.has_candidate
```
[the remark this answers is not in the document]
Comment (n2) [reply egui/src/widgets/text_edit.rs:519]: Yes, has_candidate is introduced exactly for that purpose. On my test platforms(Windows 10, Ubuntu 18.04, MacOs Big Sur and iOS 14.4), when user click other place of canvas, compositionend message is send, and such reset logit is done in compositionend.

**Review by emilk — APPROVED** [review]
Thanks 🙏

**Pull request merged**